### ISSUE TITLE
Unable to connet a mosquitto using a private IP address and WSS

### ISSUE THREAD

**Issue by ogis-onishi** [mqttjs/MQTT.js#663]
I tried to connet a mosquitto from mqtt Node.js client (mqtt.js 2.7.2) and using a private IP address and WSS as follows but it fails. Am I missing something? Could you help me fix it? 

## wss (failure case)

```
"use strict"
var mqtt = require('mqtt')
var fs = require("fs")
var uri = 'wss://172.31.12.4:8443'
var opts = {
    ca: fs.readFileSync("/home/ec2-user/certs_private_ip/ca.crt.pem")
}
var client  = mqtt.connect(uri, opts)

client.on('connect', function () {
  client.subscribe('presence')
  client.publish('presence', 'Hello mqtt')
})

client.on('message', function (topic, message) {
  // message is Buffer
  console.log(message.toString())
  client.end()
})

client.on('error', function (err) {
  console.log(JSON.stringify(err))
})

client.on('close', function () {
  console.log("closed")
})
```

## mqtts (success case)

I also checked mqtts. It succeeded so there is no promlem on mosquitto conf and certificates.
```
"use strict"
var mqtt = require('mqtt')
var fs = require("fs")
var uri = 'mqtts://172.31.12.4:8883'
var opts = {
    ca: fs.readFileSync("/home/ec2-user/certs_private_ip/ca.crt.pem")
}
var client  = mqtt.connect(uri, opts)

client.on('connect', function () {
  client.subscribe('presence')
  client.publish('presence', 'Hello mqtt')
})

client.on('message', function (topic, message) {
  // message is Buffer
  console.log(message.toString())
  client.end()
})

client.on('error', function (err) {
  console.log(JSON.stringify(err))
})

client.on('close', function () {
  console.log("closed")
})
```

## mosquitto.conf

```
# Plain MQTT protocol
listener 1883

# End of plain MQTT configuration

# MQTT over TLS/SSL
listener 8883
cafile   /home/ec2-user/certs_private_ip/ca.crt.pem
certfile /home/ec2-user/certs_private_ip/server.crt.pem
keyfile  /home/ec2-user/certs_private_ip/server.key.pem

# End of MQTT over TLS/SLL configuration

# Plain WebSockets configuration
listener 8080
protocol websockets

# End of plain Websockets configuration

# WebSockets over TLS/SSL
listener 8443
protocol websockets
cafile   /home/ec2-user/certs_private_ip/ca.crt.pem
certfile /home/ec2-user/certs_private_ip/server.crt.pem
keyfile  /home/ec2-user/certs_private_ip/server.key.pem
```

**Comment by mcollina** [comment]
It fails with what error?

**Comment by mcollina** [comment]
Does it work without TLS for websockets?

**Comment by ogis-onishi** [comment]
It succeeded without TLS (MQTT and Websocket).

**Comment by ogis-onishi** [comment]
I set the following callback but it did not throw any erros. It just called close callback.

```
client.on('error', function (err) {
  console.log(JSON.stringify(err))
})
```

**Comment by mcollina** [comment]
Can you please confirm with the latest version?

**Comment by itavy** [comment]
Try console.log(err) without json.stringify

**Comment by ogis-onishi** [comment]
Sorry. The code with MQTTS throws the error.

```
error = {"reason":"IP: 172.31.12.4 is not in the cert's list: ","host":"172.31.12.4","cert":{"subject":{"C":"JP","ST":"Tokyo","L":"Minato","O":"My Company","OU":"ITdept.","CN":"172.31.12.4"},"issuer":{"C":"JP","ST":"Tokyo","L":"Minato","O":"My Company","OU":"ITDept.","CN":"MyCA"},"subjectaltname":"DNS:172.31.12.4",
```

**Comment by mcollina** [comment]
I think you have an issue with your cert.

**Comment by itavy** [comment]
Add that ip to server cert cname

**Comment by ogis-onishi** [comment]
Without JSON.stringify

```
error = Error: Hostname/IP doesn't match certificate's altnames: "IP: 172.31.12.4 is not in the cert's list: "
```

**Comment by itavy** [comment]
See https://stackoverflow.com/a/37063612

Certificate issued for ip and not domain name works only for private ip.

**Comment by ogis-onishi** [comment]
I signed the sertificate as follows. 

```
hostname=172.31.12.4
openssl req -new -sha256 -key $1/server.key.pem -subj "/C=JP/ST=Tokyo/L=Minato/O=My Company/OU=ITdept./CN=$hostname" > $1/server.csr.pem
```

And I also added alternative subject names.
https://stackoverflow.com/questions/43665243/chrome-invalid-self-signed-ssl-cert-subject-alternative-name-missing

```
authorityKeyIdentifier=keyid,issuer
basicConstraints=CA:FALSE
keyUsage = digitalSignature, nonRepudiation, keyEncipherment, dataEncipherment
subjectAltName = @alt_names

[alt_names]
DNS.1 = 172.31.12.4
```

So, the private address is in the cert's list. But it would be OK for DNS but not good for IP address because the error for MQTTS says :  `error = {"reason":"IP: 172.31.12.4 is not in the cert's list: ","host":"172.31.12.4",
"cert":{"subject":{"C":"JP","ST":"Tokyo","L":"Minato","O":"My Company","OU":"ITdept.",
"CN":"172.31.12.4"},
"issuer":{"C":"JP","ST":"Tokyo","L":"Minato","O":"My Company","OU":"ITDept.","CN":"MyCA"},
"subjectaltname":"DNS:172.31.12.4",`
